Simple KCFG-to-KCFG transformations for cleaner summarisation

PetarMax opened the pull request:
This PR implements some simple KCFG-to-KCFG transformations, the purpose of which is to enable cleaner summarisation. The currently implemented transformations are:

- lifting of edges: `A --M steps--> B --N steps--> C` becomes `A --(M + N) steps--> C`
- lifting of splits: `A --M steps--> B --[cond_1, ..., cond_N]--> [C_1, ..., C_N]` becomes `A --[cond_1, ..., cond_N]--> [A #And cond_1 --M steps--> C_1, ..., A #And cond_N --M steps--> C_N]`

ehildenb commented on src/pyk/kcfg/kcfg.py:
```diff
@@ -854,6 +856,80 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
+            `A --M steps--> B --N steps--> C` becomes `A --(M + N) steps--> C`. Node `B` is removed.
+            The edges in question are assumed to be in place, otherwise an exception is thrown.
+        """
+        # Obtain edge `A -> B`
+        edges_to_b = self.edges(target_id=id)
+        assert len(edges_to_b) == 1
+        a_to_b = edges_to_b[0]
+
+        # Obtain edge `B -> C`
+        edges_from_b = self.edges(source_id=id)
+        assert len(edges_from_b) == 1
+        b_to_c = edges_from_b[0]
```
```suggestion
        # Obtain edge `A -> B`, `B -> C`
        a_to_b = single(self.edges(target_id=id))
        b_to_c = single(self.edges(source_id=id))
```

PetarMax replied:
Ah yes, forgot about `sngle`, thank you!

ehildenb commented on src/pyk/kcfg/kcfg.py:
```diff
@@ -854,6 +856,80 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
+        # Remove the node `B`, effectively removing the entire initial structure
+        self.remove_node(id)
+
+        # TODO: What is the correct way of handling the rules?
+        self.create_edge(a_to_b.source.id, b_to_c.target.id, a_to_b.depth + b_to_c.depth, a_to_b.rules + b_to_c.rules)
```
This is correct, the rules are what the backend reports as the rules that get you from the source node to the target node of the edge.

ehildenb commented on src/pyk/kcfg/kcfg.py:
```diff
@@ -854,6 +856,80 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
+            Node `B` is removed. If any of the `cond_i` contain variables not present in `A`, an exeception is thrown.
+            The structure in question is assumed to be in place, otherwise an exception is thrown.
+        """
+        # Obtain edge `A -> B`
+        edges_to_b = self.edges(target_id=id)
+        assert len(edges_to_b) == 1
+        a_to_b = edges_to_b[0]
+        a = a_to_b.source
+
+        # Obtain split targets`[cond_I, C_I | I = 1..N ]`
+        splits_from_b = self.splits(source_id=id)
+        assert len(splits_from_b) == 1
+        ci, substs = list(splits_from_b[0].splits.keys()), list(splits_from_b[0].splits.values())
```
```suggestion
        # Obtain edge `A -> B`, split `[ cond_I, C_I | I = 1..N ]`
        a_to_b = single(self.edges(target_id=id))
        a = a_to_b.source
        split_from_b = single(self.splits(source_id=id))
        ci, substs = list(split_from_b.splits.keys()), list(split_from_b.splits.values())
```

ehildenb replied:
`single` can be imported from `..utils`

ehildenb commented on src/pyk/kcfg/kcfg.py:
```diff
@@ -854,6 +856,80 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
+        # a `CTerm` with the configuration of `A` and the constraints of `cond_I`.
+        ai: list[NodeIdLike] = [self.create_node(CTerm(a.cterm.config, subst.constraints)).id for subst in substs]
+
+        # Create the edges `[A #And cond_1 --M steps--> C_1, ..., A #And cond_N --M steps--> C_N]`
+        # TODO: What is the correct way of handling the rules?
+        for i in range(len(ai)):
+            self.create_edge(ai[i], ci[i], a_to_b.depth, a_to_b.rules)
```
This is still the correct way. The rules list should get us from the source node to teh target node, which will be the case here.

ehildenb commented:
It looks good overall. I would:

- Add tests.
- Perhaps expose three routines that are more automated which:
  - Look for any back-to-back edges and call `lift_edge` until there are no more back-to-back edges.
  - Look for any edges into splits and call `lift_split` on them until there are no more edges into splits.

Then `kontrol minimize-summary ...` can call the first routine and the second routine and write the proof to disk again.

PetarMax commented:
On @tothtamas28's advice, I've rebased against `origin/master` to clear up the commits that were showing up but were not related to the PR, hence the force-push.

ehildenb commented on src/pyk/kcfg/kcfg.py:
```diff
@@ -854,6 +861,153 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
+        a_to_b = single(self.edges(target_id=id))
+        a = a_to_b.source
+        split_from_b = single(self.splits(source_id=id))
+        ci, substs = list(split_from_b.splits.keys()), list(split_from_b.splits.values())
```
```suggestion
        ci, csubsts = list(split_from_b.splits.keys()), list(split_from_b.splits.values())
```

These are `CSubst`, not `Subst`, so let's name them as such. https://github.com/runtimeverification/pyk/blob/af2567586fcc13163cb8aea695c807deffb1ebb7/src/pyk/cterm/cterm.py#L300

ehildenb commented:
Let's also add a test of an `APRProof` being minimized appropriately. That can go in integration-tests, you can probably use `simple-proofs.k` (`./src/tests/integration/proof/test_simple.py`) to make a KCFG happen during a proof, call `APRProof.minimize_kcfg()`, and then show the output.

ehildenb commented:
Even better would be to add it to the APRProof unit testing harness: https://github.com/runtimeverification/pyk/blob/master/src/tests/unit/test_proof.py. Make sure it exercises the case where a terminal node was in there before, and now a new one is added (the bug discovered in Kontrol integration).

PetarMax commented:
I've tested these changes with Kontrol, posting KCFGs below, the modules are evident. In summary, the KCFGs are as we would expect, the splits are being pulled up as high as possible and there are no consecutive edges. I can also confirm that the resulting rules are used correctly in CSE (with `--break-on-calls`, though, until [this PR](https://github.com/runtimeverification/hs-backend-booster/pull/544) is upstreamed). The exponential branching that was happening in `ArithmeticCallTest.test_double_add_sub_external` due to undesirable expressions in the LHS of rules is also gone.

There is an issue revealed by `ArithmeticCallTest.test_double_add_sub_external` in that a terminal node (e.g., the leaf node of a `setUp` function of a Foundry test) might be removed from the KCFG without the set of `terminal` nodes being updated, since that set is handled one level above, in `KCFGExploration`.

## `ArithmeticContract.add`

### Without minimisation:
```
┌─ 1 (root, init)
│   k: #execute ~> CONTINUATION:K
│   pc: 0
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/ArithmeticContract.sol:3:25
│
│  (341 steps)
├─ 3 (split)
│   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
│   pc: 562
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
┃
┃ (branch)
┣━━┓ constraint: VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int )
┃  │
┃  ├─ 4
┃  │   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
┃  │   pc: 562
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │
┃  │  (129 steps)
┃  ├─ 6 (terminal)
┃  │   k: #halt ~> CONTINUATION:K
┃  │   pc: 128
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: EVMC_SUCCESS
┃  │
┃  ┊  constraint: OMITTED CONSTRAINT
┃  ┊  subst: OMITTED SUBST
┃  └─ 2 (leaf, target, terminal)
┃      k: #halt ~> CONTINUATION:K
┃      pc: PC_CELL_5d410f2a:Int
┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃      statusCode: STATUSCODE_FINAL:StatusCode
┃
┗━━┓ constraint: ( notBool VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int ) )
   │
   ├─ 5
   │   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
   │   pc: 562
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: STATUSCODE:StatusCode
   │
   │  (66 steps)
   ├─ 7 (terminal)
   │   k: #halt ~> CONTINUATION:K
   │   pc: 550
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: EVMC_REVERT
   │
   ┊  constraint: OMITTED CONSTRAINT
   ┊  subst: OMITTED SUBST
   └─ 2 (leaf, target, terminal)
       k: #halt ~> CONTINUATION:K
       pc: PC_CELL_5d410f2a:Int
       callDepth: CALLDEPTH_CELL_5d410f2a:Int
       statusCode: STATUSCODE_FINAL:StatusCode
```

### With minimisation:
```
┌─ 1 (root, split, init)
│   k: #execute ~> CONTINUATION:K
│   pc: 0
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/ArithmeticContract.sol:3:25
┃
┃ (branch)
┣━━┓ constraint: VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int )
┃  │
┃  ├─ 8
┃  │   k: #execute ~> CONTINUATION:K
┃  │   pc: 0
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │   src: src/ArithmeticContract.sol:3:25
┃  │
┃  │  (470 steps)
┃  ├─ 6 (terminal)
┃  │   k: #halt ~> CONTINUATION:K
┃  │   pc: 128
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: EVMC_SUCCESS
┃  │
┃  ┊  constraint: OMITTED CONSTRAINT
┃  ┊  subst: OMITTED SUBST
┃  └─ 2 (leaf, target, terminal)
┃      k: #halt ~> CONTINUATION:K
┃      pc: PC_CELL_5d410f2a:Int
┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃      statusCode: STATUSCODE_FINAL:StatusCode
┃
┗━━┓ constraint: ( notBool VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int ) )
   │
   ├─ 9
   │   k: #execute ~> CONTINUATION:K
   │   pc: 0
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: STATUSCODE:StatusCode
   │   src: src/ArithmeticContract.sol:3:25
   │
   │  (407 steps)
   ├─ 7 (terminal)
   │   k: #halt ~> CONTINUATION:K
   │   pc: 550
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: EVMC_REVERT
   │
   ┊  constraint: OMITTED CONSTRAINT
   ┊  subst: OMITTED SUBST
   └─ 2 (leaf, target, terminal)
       k: #halt ~> CONTINUATION:K
       pc: PC_CELL_5d410f2a:Int
       callDepth: CALLDEPTH_CELL_5d410f2a:Int
       statusCode: STATUSCODE_FINAL:StatusCode
```

## `ArithmeticContract.add_sub_external`

### Without minimisation
```
┌─ 1 (root, init)
│   k: #execute ~> CONTINUATION:K
│   pc: 0
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/ArithmeticContract.sol:3:25
│
│  (463 steps)
├─ 3 (split)
│   k: #checkDepthExceeded ~> #call CONTRACT_ID:Int CONTRACT_ID:Int CONTRACT_ID:Int 0 0 ...
│   pc: 279
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
┃
┃ (branch)
┣━━┓ constraint: CALLDEPTH_CELL:Int <Int 1024
┃  │
┃  ├─ 4
┃  │   k: #checkDepthExceeded ~> #call CONTRACT_ID:Int CONTRACT_ID:Int CONTRACT_ID:Int 0 0 ...
┃  │   pc: 279
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │
┃  │  (357 steps)
┃  ├─ 6 (split)
┃  │   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
┃  │   pc: 562
┃  │   callDepth: ( CALLDEPTH_CELL:Int +Int 1 )
┃  │   statusCode: STATUSCODE:StatusCode
┃  ┃
┃  ┃ (branch)
┃  ┣━━┓ constraint: VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int )
┃  ┃  │
┃  ┃  ├─ 8
┃  ┃  │   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
┃  ┃  │   pc: 562
┃  ┃  │   callDepth: ( CALLDEPTH_CELL:Int +Int 1 )
┃  ┃  │   statusCode: STATUSCODE:StatusCode
┃  ┃  │
┃  ┃  │  (299 steps)
┃  ┃  ├─ 10 (split)
┃  ┃  │   k: JUMPI 618 bool2Word ( VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_1 ...
┃  ┃  │   pc: 610
┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  │   statusCode: EVMC_SUCCESS
┃  ┃  ┃
┃  ┃  ┃ (branch)
┃  ┃  ┣━━┓ constraint: VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int )
┃  ┃  ┃  │
┃  ┃  ┃  ├─ 12
┃  ┃  ┃  │   k: JUMPI 618 bool2Word ( VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_1 ...
┃  ┃  ┃  │   pc: 610
┃  ┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  ┃  │   statusCode: EVMC_SUCCESS
┃  ┃  ┃  │
┃  ┃  ┃  │  (143 steps)
┃  ┃  ┃  ├─ 14 (terminal)
┃  ┃  ┃  │   k: #halt ~> CONTINUATION:K
┃  ┃  ┃  │   pc: 128
┃  ┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  ┃  │   statusCode: EVMC_SUCCESS
┃  ┃  ┃  │
┃  ┃  ┃  ┊  constraint: OMITTED CONSTRAINT
┃  ┃  ┃  ┊  subst: OMITTED SUBST
┃  ┃  ┃  └─ 2 (leaf, target, terminal)
┃  ┃  ┃      k: #halt ~> CONTINUATION:K
┃  ┃  ┃      pc: PC_CELL_5d410f2a:Int
┃  ┃  ┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃  ┃  ┃      statusCode: STATUSCODE_FINAL:StatusCode
┃  ┃  ┃
┃  ┃  ┗━━┓ constraint: ( notBool VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int ) )
┃  ┃     │
┃  ┃     ├─ 13
┃  ┃     │   k: JUMPI 618 bool2Word ( VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_1 ...
┃  ┃     │   pc: 610
┃  ┃     │   callDepth: CALLDEPTH_CELL:Int
┃  ┃     │   statusCode: EVMC_SUCCESS
┃  ┃     │
┃  ┃     │  (66 steps)
┃  ┃     ├─ 15 (terminal)
┃  ┃     │   k: #halt ~> CONTINUATION:K
┃  ┃     │   pc: 550
┃  ┃     │   callDepth: CALLDEPTH_CELL:Int
┃  ┃     │   statusCode: EVMC_REVERT
┃  ┃     │
┃  ┃     ┊  constraint: OMITTED CONSTRAINT
┃  ┃     ┊  subst: OMITTED SUBST
┃  ┃     └─ 2 (leaf, target, terminal)
┃  ┃         k: #halt ~> CONTINUATION:K
┃  ┃         pc: PC_CELL_5d410f2a:Int
┃  ┃         callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃  ┃         statusCode: STATUSCODE_FINAL:StatusCode
┃  ┃
┃  ┗━━┓ constraint: ( notBool VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int ) )
┃     │
┃     ├─ 9
┃     │   k: JUMPI 570 bool2Word ( VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705: ...
┃     │   pc: 562
┃     │   callDepth: ( CALLDEPTH_CELL:Int +Int 1 )
┃     │   statusCode: STATUSCODE:StatusCode
┃     │
┃     │  (73 steps)
┃     ├─ 11 (terminal)
┃     │   k: #halt ~> CONTINUATION:K
┃     │   pc: 295
┃     │   callDepth: CALLDEPTH_CELL:Int
┃     │   statusCode: EVMC_REVERT
┃     │
┃     ┊  constraint: OMITTED CONSTRAINT
┃     ┊  subst: OMITTED SUBST
┃     └─ 2 (leaf, target, terminal)
┃         k: #halt ~> CONTINUATION:K
┃         pc: PC_CELL_5d410f2a:Int
┃         callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃         statusCode: STATUSCODE_FINAL:StatusCode
┃
┗━━┓ constraint: 1024 <=Int CALLDEPTH_CELL:Int
   │
   ├─ 5
   │   k: #checkDepthExceeded ~> #call CONTRACT_ID:Int CONTRACT_ID:Int CONTRACT_ID:Int 0 0 ...
   │   pc: 279
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: STATUSCODE:StatusCode
   │
   │  (76 steps)
   ├─ 7 (terminal)
   │   k: #halt ~> CONTINUATION:K
   │   pc: 295
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: EVMC_REVERT
   │
   ┊  constraint: OMITTED CONSTRAINT
   ┊  subst: OMITTED SUBST
   └─ 2 (leaf, target, terminal)
       k: #halt ~> CONTINUATION:K
       pc: PC_CELL_5d410f2a:Int
       callDepth: CALLDEPTH_CELL_5d410f2a:Int
       statusCode: STATUSCODE_FINAL:StatusCode
```

### With minimisation
```
┌─ 1 (root, split, init)
│   k: #execute ~> CONTINUATION:K
│   pc: 0
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/ArithmeticContract.sol:3:25
┃
┃ (branch)
┣━━┓ constraint: CALLDEPTH_CELL:Int <Int 1024
┃  │
┃  ├─ 35 (split)
┃  │   k: #execute ~> CONTINUATION:K
┃  │   pc: 0
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │   src: src/ArithmeticContract.sol:3:25
┃  ┃
┃  ┃ (branch)
┃  ┣━━┓ constraint: ( maxUInt256 -Int VV1_y_114b9705:Int ) <Int VV0_x_114b9705:Int
┃  ┃  │
┃  ┃  ├─ 41
┃  ┃  │   k: #execute ~> CONTINUATION:K
┃  ┃  │   pc: 0
┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  │   statusCode: STATUSCODE:StatusCode
┃  ┃  │   src: src/ArithmeticContract.sol:3:25
┃  ┃  │
┃  ┃  │  (552 steps)
┃  ┃  ├─ 28 (terminal)
┃  ┃  │   k: #halt ~> CONTINUATION:K
┃  ┃  │   pc: 295
┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  │   statusCode: EVMC_REVERT
┃  ┃  │
┃  ┃  ┊  constraint: OMITTED CONSTRAINT
┃  ┃  ┊  subst: OMITTED SUBST
┃  ┃  └─ 2 (leaf, target, terminal)
┃  ┃      k: #halt ~> CONTINUATION:K
┃  ┃      pc: PC_CELL_5d410f2a:Int
┃  ┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃  ┃      statusCode: STATUSCODE_FINAL:StatusCode
┃  ┃
┃  ┗━━┓ constraint: VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int )
┃     │
┃     ├─ 42 (split)
┃     │   k: #execute ~> CONTINUATION:K
┃     │   pc: 0
┃     │   callDepth: CALLDEPTH_CELL:Int
┃     │   statusCode: STATUSCODE:StatusCode
┃     │   src: src/ArithmeticContract.sol:3:25
┃     ┃
┃     ┃ (branch)
┃     ┣━━┓ constraint: VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int )
┃     ┃  │
┃     ┃  ├─ 45
┃     ┃  │   k: #execute ~> CONTINUATION:K
┃     ┃  │   pc: 0
┃     ┃  │   callDepth: CALLDEPTH_CELL:Int
┃     ┃  │   statusCode: STATUSCODE:StatusCode
┃     ┃  │   src: src/ArithmeticContract.sol:3:25
┃     ┃  │
┃     ┃  │  (921 steps)
┃     ┃  ├─ 33 (terminal)
┃     ┃  │   k: #halt ~> CONTINUATION:K
┃     ┃  │   pc: 128
┃     ┃  │   callDepth: CALLDEPTH_CELL:Int
┃     ┃  │   statusCode: EVMC_SUCCESS
┃     ┃  │
┃     ┃  ┊  constraint: OMITTED CONSTRAINT
┃     ┃  ┊  subst: OMITTED SUBST
┃     ┃  └─ 2 (leaf, target, terminal)
┃     ┃      k: #halt ~> CONTINUATION:K
┃     ┃      pc: PC_CELL_5d410f2a:Int
┃     ┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃     ┃      statusCode: STATUSCODE_FINAL:StatusCode
┃     ┃
┃     ┗━━┓ constraint: ( notBool VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int ) )
┃        │
┃        ├─ 46
┃        │   k: #execute ~> CONTINUATION:K
┃        │   pc: 0
┃        │   callDepth: CALLDEPTH_CELL:Int
┃        │   statusCode: STATUSCODE:StatusCode
┃        │   src: src/ArithmeticContract.sol:3:25
┃        │
┃        │  (844 steps)
┃        ├─ 34 (terminal)
┃        │   k: #halt ~> CONTINUATION:K
┃        │   pc: 550
┃        │   callDepth: CALLDEPTH_CELL:Int
┃        │   statusCode: EVMC_REVERT
┃        │
┃        ┊  constraint: OMITTED CONSTRAINT
┃        ┊  subst: OMITTED SUBST
┃        └─ 2 (leaf, target, terminal)
┃            k: #halt ~> CONTINUATION:K
┃            pc: PC_CELL_5d410f2a:Int
┃            callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃            statusCode: STATUSCODE_FINAL:StatusCode
┃
┗━━┓ constraint: 1024 <=Int CALLDEPTH_CELL:Int
   │
   ├─ 36
   │   k: #execute ~> CONTINUATION:K
   │   pc: 0
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: STATUSCODE:StatusCode
   │   src: src/ArithmeticContract.sol:3:25
   │
   │  (539 steps)
   ├─ 27 (terminal)
   │   k: #halt ~> CONTINUATION:K
   │   pc: 295
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: EVMC_REVERT
   │
   ┊  constraint: OMITTED CONSTRAINT
   ┊  subst: OMITTED SUBST
   └─ 2 (leaf, target, terminal)
       k: #halt ~> CONTINUATION:K
       pc: PC_CELL_5d410f2a:Int
       callDepth: CALLDEPTH_CELL_5d410f2a:Int
       statusCode: STATUSCODE_FINAL:StatusCode
```

PetarMax commented:
When it comes to`ArithmeticCallTest.test_double_add_sub_external`, the KCFG (see below) is much better, it does not branch spuriously anymore and the proof finishes in 5 minutes. It will be faster when the booster unification fix is merged, since the proof is now run with `--break-on-calls`. 

```
┌─ 1 (root, split, init)
│   k: #execute ~> CONTINUATION:K
│   pc: 0
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/ArithmeticContract.sol:3:25
┃
┃ (branch)
┣━━┓ constraint: CALLDEPTH_CELL:Int <Int 1024
┃  │
┃  ├─ 35 (split)
┃  │   k: #execute ~> CONTINUATION:K
┃  │   pc: 0
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │   src: src/ArithmeticContract.sol:3:25
┃  ┃
┃  ┃ (branch)
┃  ┣━━┓ constraint: ( maxUInt256 -Int VV1_y_114b9705:Int ) <Int VV0_x_114b9705:Int
┃  ┃  │
┃  ┃  ├─ 41
┃  ┃  │   k: #execute ~> CONTINUATION:K
┃  ┃  │   pc: 0
┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  │   statusCode: STATUSCODE:StatusCode
┃  ┃  │   src: src/ArithmeticContract.sol:3:25
┃  ┃  │
┃  ┃  │  (552 steps)
┃  ┃  ├─ 28 (terminal)
┃  ┃  │   k: #halt ~> CONTINUATION:K
┃  ┃  │   pc: 295
┃  ┃  │   callDepth: CALLDEPTH_CELL:Int
┃  ┃  │   statusCode: EVMC_REVERT
┃  ┃  │
┃  ┃  ┊  constraint: OMITTED CONSTRAINT
┃  ┃  ┊  subst: OMITTED SUBST
┃  ┃  └─ 2 (leaf, target, terminal)
┃  ┃      k: #halt ~> CONTINUATION:K
┃  ┃      pc: PC_CELL_5d410f2a:Int
┃  ┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃  ┃      statusCode: STATUSCODE_FINAL:StatusCode
┃  ┃
┃  ┗━━┓ constraint: VV0_x_114b9705:Int <=Int ( maxUInt256 -Int VV1_y_114b9705:Int )
┃     │
┃     ├─ 42 (split)
┃     │   k: #execute ~> CONTINUATION:K
┃     │   pc: 0
┃     │   callDepth: CALLDEPTH_CELL:Int
┃     │   statusCode: STATUSCODE:StatusCode
┃     │   src: src/ArithmeticContract.sol:3:25
┃     ┃
┃     ┃ (branch)
┃     ┣━━┓ constraint: VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int )
┃     ┃  │
┃     ┃  ├─ 45
┃     ┃  │   k: #execute ~> CONTINUATION:K
┃     ┃  │   pc: 0
┃     ┃  │   callDepth: CALLDEPTH_CELL:Int
┃     ┃  │   statusCode: STATUSCODE:StatusCode
┃     ┃  │   src: src/ArithmeticContract.sol:3:25
┃     ┃  │
┃     ┃  │  (921 steps)
┃     ┃  ├─ 33 (terminal)
┃     ┃  │   k: #halt ~> CONTINUATION:K
┃     ┃  │   pc: 128
┃     ┃  │   callDepth: CALLDEPTH_CELL:Int
┃     ┃  │   statusCode: EVMC_SUCCESS
┃     ┃  │
┃     ┃  ┊  constraint: OMITTED CONSTRAINT
┃     ┃  ┊  subst: OMITTED SUBST
┃     ┃  └─ 2 (leaf, target, terminal)
┃     ┃      k: #halt ~> CONTINUATION:K
┃     ┃      pc: PC_CELL_5d410f2a:Int
┃     ┃      callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃     ┃      statusCode: STATUSCODE_FINAL:StatusCode
┃     ┃
┃     ┗━━┓ constraint: ( notBool VV2_z_114b9705:Int <=Int ( VV0_x_114b9705:Int +Int VV1_y_114b9705:Int ) )
┃        │
┃        ├─ 46
┃        │   k: #execute ~> CONTINUATION:K
┃        │   pc: 0
┃        │   callDepth: CALLDEPTH_CELL:Int
┃        │   statusCode: STATUSCODE:StatusCode
┃        │   src: src/ArithmeticContract.sol:3:25
┃        │
┃        │  (844 steps)
┃        ├─ 34 (terminal)
┃        │   k: #halt ~> CONTINUATION:K
┃        │   pc: 550
┃        │   callDepth: CALLDEPTH_CELL:Int
┃        │   statusCode: EVMC_REVERT
┃        │
┃        ┊  constraint: OMITTED CONSTRAINT
┃        ┊  subst: OMITTED SUBST
┃        └─ 2 (leaf, target, terminal)
┃            k: #halt ~> CONTINUATION:K
┃            pc: PC_CELL_5d410f2a:Int
┃            callDepth: CALLDEPTH_CELL_5d410f2a:Int
┃            statusCode: STATUSCODE_FINAL:StatusCode
┃
┗━━┓ constraint: 1024 <=Int CALLDEPTH_CELL:Int
   │
   ├─ 36
   │   k: #execute ~> CONTINUATION:K
   │   pc: 0
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: STATUSCODE:StatusCode
   │   src: src/ArithmeticContract.sol:3:25
   │
   │  (539 steps)
   ├─ 27 (terminal)
   │   k: #halt ~> CONTINUATION:K
   │   pc: 295
   │   callDepth: CALLDEPTH_CELL:Int
   │   statusCode: EVMC_REVERT
   │
   ┊  constraint: OMITTED CONSTRAINT
   ┊  subst: OMITTED SUBST
   └─ 2 (leaf, target, terminal)
       k: #halt ~> CONTINUATION:K
       pc: PC_CELL_5d410f2a:Int
       callDepth: CALLDEPTH_CELL_5d410f2a:Int
       statusCode: STATUSCODE_FINAL:StatusCode
```

What is different from the other two proofs is that we do have `NDBranches` from using the summary of `ArithmeticCallTest.test_double_add_sub_external`, but these are not really non-deterministic in the sense that their path conditions are disjoint and that they could be characterised as 3-way split, with `notBool A`, `A andBool notBool B` and `A andBool B`. This could be the next step of these transformations, together with some sort of joining of 2-way splits into multi-way splits.

ehildenb commented on src/tests/unit/test_kcfg.py:
```diff
@@ -441,6 +441,284 @@ def test_aliases() -> None:
 
 
+def minimisation_test_kcfg() -> KCFG:
+    #                                               25   /-- Y >=Int 0 --> 10
+    #     5    10    15    20   /-- X >=Int 0 --> 6 --> 8
+    #  1 --> 2 --> 3 --> 4 --> 5                         \-- Y  <Int 0 --> 11
+    #                           \                   30    35     40
+    #                            \-- X <Int 0 --> 7 --> 9 --> 12 --> 13
```
I think this KCFG can be represented as follows (if you thread the option `with_cond` through to `node_dicts` from `node`):

```
    d = {
        'nodes': node_dicts(13, with_cond=True),
        'edges': edge_dicts((1,2),(2,3),(3,4),(4,5),(6,8),(7,9),(9,12),(12,13)),
        'splits': [split(5,[6,7]), split(8,[10,11])],
    }
    cfg = KCFG.from_dict(d)
```

You could also thread through teh option to specificy the depth/rules of each edge, but I don't think that it would add much to the test itself to do so.

PetarMax replied:
I have rewritten the tests to use as many existing constructs as possible, and they are cleaner. I have also done the following:
- factored out `x_equals` since it is used in multiple places;
- changed `KVariable('x')` to `KVariable('X')` since variables can't start with a lowercase letter;
- factored out `to_subst` since it is used in multiple places; and
- extended `_make_edge_dict` in `edge_dicts` to also handle `rules` because the functionality I am testing manipulates edge rules.

ehildenb commented:
I'm not a huge fan of how much vertical space the comments and code is taking up, lots of newlines for just a few lines of code. So I'd suggest compacting them, it makes it more in line with the rest of the file and codestyle elsewhere.

ehildenb commented:
Looks good overall, let's just use the existing infrastructure for KCFG testing instead of a new approach, and compact the code a bit.

PetarMax commented:
> I'm not a huge fan of how much vertical space the comments and code is taking up, lots of newlines for just a few lines of code. So I'd suggest compacting them, it makes it more in line with the rest of the file and codestyle elsewhere.

I have now compacted the code in `kcfg.py`. I've left the inlined comments (`#`) but I am happy to remove them if you think they don't bring value.

PetarMax commented:
I've just realised that the substutitions carried in the splits are not updated by the lifting. For example, the transformation goes from
```
    #                                                     25    /-- X >=Int 5 --> 20
    #      5     10     15     20    /-- X >=Int 0 --> 16 --> 18
    #  11 --> 12 --> 13 --> 14 --> 15                           \-- X  <Int 5 --> 21
    #                               \                     30     35     40
    #                                \-- X  <Int 0 --> 17 --> 19 --> 22 --> 23
```
to
```
    #                                             75
    #                        /-- Y >=Int 0 --> 28 --> 20
    #    /-- X >=Int 0 --> 24                     75
    #  11                    \-- Y  <Int 0 --> 29 --> 21
    #    \                    155
    #     \-- X <Int 0 --> 25 --> 23
```
but the split that moved from `15` to `11` still carries the substitution from `15` to `16` and `15` to `17` instead of from `11` to `24` and `11` to `25`.

Should these substitutions also be updated?

PetarMax commented:
> Should these substitutions also be updated?

I've examined the codebase, and I found the following:
- in `APRProof.path_constraints`, information from the `subst` part of `CSubst` is used to create the path constraint;
- in `APRProof.construct_node_refutation`, it is expected that the `subst` of a `Split` is empty;
- in `Split.replace_target`, when the node is replaced the `subst` is not updated.

ehildenb commented:
Let's start a new PR with a test that exercises that issue and fix it there!
